Library/Homebrew/utils/pypi.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+module PyPI
+  module_function
+
+  # Get name, url, and version for a given pypi package
+  def get_pypi_info(package, version)
+    metadata_url = "https://pypi.org/pypi/#{package}/#{version}/json"
+    out, _, status = curl_output metadata_url, "--location"
+
+    return unless status.success?
+
+    begin
+      json = JSON.parse out
+    rescue JSON::ParserError
+      return
+    end
+
+    sdist = json["urls"].find { |url| url["packagetype"] == "sdist" }
+    [json["info"]["name"], sdist["url"], sdist["digests"]["sha256"]]
+  end
+
+  def update_python_resources!(formula, constraints = nil, print_only: false, silent: false,
+                               ignore_non_pypi_packages: false)
+    # PyPI package name isn't always the same as the formula name. Try to infer from the URL.
+    pypi_name = if formula.stable.url.start_with?("https://files.pythonhosted.org/packages/")
+      File.basename(formula.stable.url).match(/^(.+)-[a-z0-9.]+$/)[1]
+    else
+      formula.name
+    end
+
+    # See if package is a PyPI package
+    if get_pypi_info(pypi_name, formula.version).nil?
+      odie "\"#{pypi_name}\" is not a PyPI package." unless ignore_non_pypi_packages
+      return
+    end
+
+    # Ensure there are no non-pypi resources when not running with --print-only
+    non_pypi_resources = formula.resources.reject do |resource|
+      resource.url.start_with? "https://files.pythonhosted.org/packages/"
+    end
+    if !non_pypi_resources.empty? && !print_only
+      odie "\"#{formula.name}\" contains non-PyPI resources. Please update the resources manually."
+    end
+
+    # Get list of required packages
+    ohai "Retrieving PyPI dependencies for \"#{pypi_name}==#{formula.version}\"" unless print_only || silent
+    args = %w[--json --no-cache-dir]
+    args << constraints unless constraints.nil?
```

Comment:
From the flag description/readme, I was wondering how I would pass multiple constraints. It looks like here they are unpacked as separate args, so would I then do eg `--constraints='cffi>1.11.3,<1.13 requests~=1.1'` and it's parsed based on the whitespaces?

Reply:
No, `'cffi>1.11.3,<1.13 requests~=1.1'` would be interpreted as one argument. Essentially, this is what would be called:

```bash
pipgrip foo==1.2.3 --json --no-cache-dir 'cffi>1.11.3,<1.13 requests~=1.1'
```

If the `--constraints` flag is desired, the logic should probably be rethought to allow multiple arguments. Thanks for bringing this up 👍